File: Scala/src/main/scala/AlgorithmicHeights/PAR.scala
package AlgorithmicHeights

object PAR {
  import scala.annotation.tailrec

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    val elem: Int = array(ix)
    array(ix) = array(jy)
    array(jy) = elem
  }

  def twoWayPartitioning(array: Array[Int], pivot: Int, n: Int): Unit = {
    @tailrec
    def loop(startIx: Int, endIx: Int): Unit = {
      if (startIx > endIx) ()
      else if (array(startIx) <= pivot) {
        swapElems(array, startIx, startIx - 1)
        loop(startIx + 1, endIx)
       }
      else {
        swapElems(array, startIx, endIx)
        loop(startIx, endIx - 1)
      }
    }

    loop(1, n - 1)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    twoWayPartitioning(array, array(0), n)
    println(array.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/par.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let two_way_partitioning (arr : int array) (pivot : int) (n : int) : unit =
    let swap i j =
        let elem = arr.(i) in
        arr.(i) <- arr.(j);
        arr.(j) <- elem in
    let rec loop start_ix end_ix =
        if start_ix > end_ix then ()
        else if arr.(start_ix) <= pivot then
            (swap start_ix (start_ix - 1);
            loop (start_ix + 1) end_ix)
        else
            (swap start_ix end_ix;
            loop start_ix (end_ix - 1)) in
    loop 1 (n - 1)


let () =
    let n = read_int () in
    let arr = convert_to_ints (read_line ()) in
    two_way_partitioning arr arr.(0) n;
    Array.(arr |> map string_of_int |> to_list) |> String.concat " " |> print_endline
